### PR TITLE
Transition `HashCodeCombiner` to use the `aspnet/Common` variation.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Updated `LocationTagged<TValue>` to handle `null` implicit values.
- Removed `InternalsVisibleTo` declaration on `Microsoft.AspNet.Razor` to the runtime test project.
- Updated `TagHelperDescriptor` tests to utilize helper methods to construct `TagHelperDescriptor`s. This was needed since the `InternalsVisibleTo` declaration was removed.
- Formatted `project.json`s in different commit.

#449

### CODE REVIEW

File: src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -5,5 +5,4 @@
 using System.Runtime.CompilerServices;
 
 [assembly: InternalsVisibleTo("Microsoft.AspNet.Razor.Test")]
-[assembly: InternalsVisibleTo("Microsoft.AspNet.Razor.Runtime.Test")]
```
Comment (anurse) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: One down! Yay :)

File: src/Microsoft.AspNet.Razor.Runtime/project.json [hunk 2]
```diff
@@ -1,36 +1,52 @@
...
-            }
-        }
+    "Microsoft.Framework.CopyOnWriteDictionary.Sources": {
+      "type": "build",
+      "version": "1.0.0-*"
+    },
+    "Microsoft.Framework.HashCodeCombiner.Sources": {
+      "type": "build",
+      "version": "1.0.0-*"
+    },
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/project.json]: What needs this? The Runtime project isn't currently using a `HashCodeCombiner`

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/project.json]: :+1:

**Comment by anurse** [comment]
:shipit:

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs [hunk 3]
```diff
@@ -717,9 +717,7 @@ public void CreateDescriptors_BuildsDescriptorsWithInheritedProperties()
         {
             // Arrange
             var errorSink = new ErrorSink();
-
-            // Also confirm isStringProperty is calculated correctly.
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs]: This was important when I added `isStringProperty`. Why isn't it still?

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs [hunk 4]
```diff
@@ -1883,6 +1856,35 @@ public void ValidateTagHelperAttributeDescriptor_WithValidPrefix_ReturnsTrue(str
             return data;
         }
 
+        private static TagHelperDescriptor CreateTagHelperDescriptor(
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs]: This refactoring removes coverage of a number of `TagHelperDescriptor` constructors. What is the upside?

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs [hunk 5]
```diff
@@ -25,9 +25,14 @@ private static TagHelperDescriptor Valid_PlainTagHelperDescriptor
...
-                                               Valid_PlainTagHelperType.FullName,
-                                               AssemblyName);
+                return new TagHelperDescriptor(
+                    prefix:  string.Empty,
+                    tagName: "valid_plain",
+                    typeName: Valid_PlainTagHelperType.FullName,
+                    assemblyName: AssemblyName,
+                    attributes: Enumerable.Empty<TagHelperAttributeDescriptor>(),
+                    requiredAttributes: Enumerable.Empty<string>(),
+                    designTimeDescriptor: null);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs]: Ick. These constructor overloads exist in part to avoid the extra verbiage in our tests.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs [hunk 6]
```diff
@@ -1402,6 +1422,13 @@ private static TagHelperDescriptor CreatePrefixedStringDescriptor(string prefix)
                 assemblyName: stringType.GetTypeInfo().Assembly.GetName().Name);
         }
 
+        private static TagHelperDirectiveDescriptor CreateTagHelperDirectiveDescriptor(
+            string directiveText,
+            TagHelperDirectiveType directiveType)
+        {
+            return new TagHelperDirectiveDescriptor(directiveText, SourceLocation.Zero, directiveType);
+        }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs]: Again, what is the upside of this refactoring?

File: src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: Hmm, what is the upside of removing this IVF? We have no restrictions against making `internal`s visible to test projects. This change is not needed at all, even if we need `HashCodeCombiner` in the test project (because the Runtime product code doesn't use that class).

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs (hunk 4 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs]: There's only 1 `TagHelperDescriptor` constructor used at runtime, the public one. The other internal ones were there for testing convenience.

File: src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: With the `InternalsVisibleTo` `Microsoft.AspNet.Razor.Runtime` would never be able to use an `aspnet/Common` package that `Microsoft.AspNet.Razor` was using. 

Currently not an issue but definitely an unneeded restriction.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs (hunk 3 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs]: This value is determined by the single public `TagHelperDescriptor` constructor. We only use the public ctor in src projects.

File: src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: Why would we *keep* an unnecessary IVT?

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: 1. @NTaylorMullen's comment just above (about using `aspnet/Common` packages) is not correct. `Microsoft.AspNet.Razor.Runtime` and `Microsoft.AspNet.Razor` are already using the same `aspnet/Common` packages e.g. `Microsoft.Framework.NotNullAttribute.Sources`. Problems online arise if a test project references both assemblies (i.e. must be the `Microsoft.AspNet.Razor.Runtime.Test` project) _and_ the test project itself uses something from the duplicate `aspnet/Common` packages. Remote case and not a problem we need to fix today.
2. Because this part of the change represents churn for churn's sake.
3. We don't make changes based on personal preferences which seems to be the real issue here.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs (hunk 3 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs]: The test was previously explicit about all properties. Now it's not. :-1:

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorFactoryTest.cs]: Right. Now things are less convenient.

In addition, why bother wrapping a single constructor call with a pseudo-helper? This method and the next one don't have a reason for being. Certainly aren't necessary with the existing `internal` constructors.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperDescriptorResolverTest.cs]: Realized this one doesn't even have an optional parameter. One argument less isn't much of an advantage over calling the constructor directly, even if you insist on not using the `internal` one.

**Comment by dougbu** [comment]
:watch:

Suggestion: If we're going to rewrite tests to avoid a few `internal` "for testing only" constructors, it should be done separately, because it's necessary or at least because we've discussed the point and agreed (grudgingly is fine of course :smiley_cat:). Further, we should completely remove the constructors; not update one of the two test projects using them to avoid a harmless IVT attribute.

File: src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: IVT is allowed only from product assembly X to unit test assembly X.Tests. This isn't personal preference.

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: @Eilon both of these IVT attributes are from this product assembly to the unit test assemblies. Are you saying the relationship is required to be 1:1?

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: The MVC repo also contains a number of product projects with IVT attributes for more than one test project. If 1:1 is a requirement, is this new?

Comment (Eilon) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: Nothing new: https://github.com/aspnet/Home/wiki/Engineering-guidelines#when-to-use-internals-vs-public-and-when-to-use-internalsvisibleto

This seems like an opportunistic fix.

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: These IVT attributes are about unit testing. Where is the 1:1 requirement in the guidelines?

This seems like an unrelated change mixed into an actual fix.

Comment (Eilon) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: It says:

>  `InternalsVisibleTo` is used only to allow a unit test to test internal types and members of its runtime assembly.

Because there is a 1:1 mapping between a unit test assembly and a runtime assembly<sup>[citation available](https://github.com/aspnet/Home/wiki/Engineering-guidelines#assembly-naming)</sup>, the IVT restriction is the logical conclusion of those two guidelines.

Let's please table this discussion now.

**Comment by NTaylorMullen** [comment]
@dougbu The internal ctors are just one of the things used by `Microsoft.AspNet.Razor.Tests`. The IVT to the `Microsoft.AspNet.Razor.Test` project is used for muchhh more than internal ctors (method calls, classes etc.). Not seeing a compelling reason to break this PR apart, it's relatively small.

File: src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Properties/AssemblyInfo.cs]: :ok:

**Comment by dougbu** [comment]
> Not seeing a compelling reason to break this PR apart

We do individual commits / PRs for individual issues. My suggestion is to at least split the commits to match the two unrelated things in this PR::
- Use the new aspnet/Common `HashCodeCombiner`
- Remove IVT from Razor to Razor.Runtime.Test

**Comment by NTaylorMullen** [comment]
Closing out this PR since the HashCodeCombiner bit was signed off on. WIll open another one for the IVT since that seems to have caused the most bit of conflict.
